modules/camera/camera_x11.cpp

```diff
@@ -0,0 +1,770 @@
+		Ref<CameraFeedX11> feed = (Ref<CameraFeedX11>)feeds[j];
+		std::string dev_name = feed->get_device()->dev_name;
+		it = std::find(devs.begin(), devs.end(), dev_name);
+		// also check if device is still openable
+		if (it == devs.end() || !feed->get_device()->check_device(!alive)) {
+			remove_feed(feed);
+		}
+	}
+
+	for (i = 0; i < devs.size(); ++i) {
+		// keep existing devices
+		// currently only check by /dev/video* name
+		bool found = false;
+		for (j = 0; j < feeds.size(); ++j) {
+			Ref<CameraFeedX11> feed = (Ref<CameraFeedX11>)feeds[j];
+			if (devs[i] == feed->get_device()->dev_name) {
+				found = true;
+				break;
+			}
+		}
+		if (!found) {
+			// create new device and check if it is compatible
+			V4l2_Device *dev = memnew(V4l2_Device(devs[i].c_str(), &this->funcs));
+			if (dev->check_device(!alive)) {
+				Ref<CameraFeedX11> newfeed;
+				newfeed.instance();
+				newfeed->set_device(dev);
+
+				// assume display camera so inverse
+				Transform2D transform = Transform2D(-1.0, 0.0, 0.0, -1.0, 1.0, 1.0);
+				newfeed->set_transform(transform);
+
+				add_feed(newfeed);
+			} else {
+				memdelete(dev);
+			}
+		}
+	}
+};
+
+void CameraX11::check_change() {
+	// simple hotplug functionality
+	// checks for new cameras (or removed cameras)
+	// every second.
+	while (alive) {
+		update_feeds();
+		usleep(1000000);
+	}
+}
+
+CameraX11::CameraX11() {
+	// determine if libv4l2 is installed and
+	// set functions appropriately
+	libv4l2 = dlopen("libv4l2.so.0", RTLD_NOW);
+
+	if (libv4l2 == NULL) {
+		// the default v4l2 functions
+		this->funcs.open = &open;
+		this->funcs.close = &close;
+		this->funcs.dup = &dup;
+		this->funcs.ioctl = &ioctl;
+		this->funcs.read = &read;
+		this->funcs.mmap = &mmap;
+		this->funcs.munmap = &munmap;
+		this->funcs.libv4l2 = false;
+#ifdef DEBUG_ENABLED
+		print_line("libv4l2.so not found. Try standard v4l2 instead.");
+#endif
+	} else {
+		// the libv4l2 functions
+		this->funcs.open = (int (*)(const char *, int, ...))dlsym(libv4l2, "v4l2_open");
+		this->funcs.close = (int (*)(int))dlsym(libv4l2, "v4l2_close");
+		this->funcs.dup = (int (*)(int))dlsym(libv4l2, "v4l2_dup");
+		this->funcs.ioctl = (int (*)(int, unsigned long int, ...))dlsym(libv4l2, "v4l2_ioctl");
+		this->funcs.read = (long int (*)(int, void *, size_t))dlsym(libv4l2, "v4l2_read");
+		this->funcs.mmap = (void *(*)(void *, size_t, int, int, int, int64_t))dlsym(libv4l2, "v4l2_mmap");
+		this->funcs.munmap = (int (*)(void *, size_t))dlsym(libv4l2, "v4l2_munmap");
+		this->funcs.libv4l2 = true;
+#ifdef DEBUG_ENABLED
+		print_line("libv4l2 found.");
+#endif
+	}
+
+	// Find available cameras we have at this time
+	update_feeds();
+
+	// start the hotplug thread
+	// alive is also used to trigger debug output
+	// (as devices that are incompatible would always be checked
+	// and therefore print always debug messages)
+	alive = true;
+	hotplug_thread = std::thread(&CameraX11::check_change, this);
+};
```

Comment:
maybe only start the thread if library was loaded successfully?

Reply:
Well I use both, the libv4l2 library (which is an LGPL wrapper above v4l2) and if it is not present, I use v4l2 (which is in the linux kernel). It should be possible to read from a camera using only v4l2 if it provides the images already in RGB format.
So there is a change the webcam would work even if libv4l2 is not installed.

Reply:
ah ok cool :)